internal/server/boltdbstate/task.go

```diff
@@ -79,6 +79,40 @@ func (s *State) TaskDelete(ref *pb.Ref_Task) error {
 	return err
 }
 
+// TaskCancel cancels a tasks jobs by task id or run job id reference.
+// NOTE(briancain): this way means each cancel is its own transaction and commit, not the greatest.
+// Previously I attempted to implement this with taskCancel where the entire job
+// triple was canceled inside a single transaction. This ended up deadlocking because
+// when we cancel a job, we also call a `db.Update` on each job when we update its
+// state in the database. This caused a deadlock. For now we cancel each job
+// separately.
+func (s *State) TaskCancel(ref *pb.Ref_Task) error {
+	task, err := s.TaskGet(ref)
+	if err != nil {
+		return err
+	}
+
+	s.log.Debug("canceling start job for task", "task id", task.Id, "start job id", task.StartJob.Id)
+	err = s.JobCancel(task.StartJob.Id, false)
```

Comment:
I haven't looked into it: but I thought we setup a dependency tree in the job system here. If you can cancel a parent, I think it automatically cancels all dependents (there is a setting to override that I think?). Should we just lean in on that and not manually cancel the triple, and instead just cancel the root (start)?

Reply:
That's true, I'll see if we're setting up the task triple as a dependency when we schedule those jobs! Canceling a single root job and letting the internal system handle downstream dependencies sounds legit.

Reply:
Looks like currently our job system will only cancel dependent jobs off the parent if they are `QUEUED` and no other state. From that standpoint for now - I think we should explicitly cancel each of the Task triples to ensure they aren't left dangling. Then perhaps next steps is figuring out how exactly we want the job system to cancel dependent jobs and if only canceling QUEUED jobs still makes sense?